### PR TITLE
[SPARK-38107][SQL][FOLLOWUP] Refine the error-class name and message for grouped agg pandas UDF

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR follow-up https://github.com/apache/spark/pull/35656 to make error-class name and error message better.

### Why are the changes needed?

The existing class name `CANNOT_USE_MIXTURE ` and its message `"Cannot use a mixture of aggregate function and group aggregate pandas UDF"` don't seem to clearly elaborate why the error is occurred.

### Does this PR introduce _any_ user-facing change?

Yes, the error message is changed from `"The group aggregate pandas UDF %s cannot be invoked in the same SELECT list as other, non-pandas aggregate functions."` to `"The group aggregate pandas UDF <pandas_udf_name> cannot be invoked in the same SELECT list as other, non-pandas aggregate functions."`

### How was this patch tested?

The existing test should cover.

### CODE REVIEW

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -22,8 +22,8 @@
   "CANNOT_UP_CAST_DATATYPE" : {
     "message" : [ "Cannot up cast %s from %s to %s.\n%s" ]
   },
-  "CANNOT_USE_MIXTURE" : {
-    "message" : [ "Cannot use a mixture of aggregate function and group aggregate pandas UDF" ]
+  "INVALID_PANDAS_UDF_PLACEMENT" : {
+    "message" : [ "The group aggregate pandas UDF %s cannot be invoked in the same SELECT list as other, non-pandas aggregate functions." ]
```
Comment (HyukjinKwon) [inline core/src/main/resources/error/error-classes.json:26]:
```suggestion
    "message" : [ "The group aggregate pandas UDF %s cannot be invoked together with other, non-pandas aggregate functions." ]
```

Since we have DataFrame API too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 2]
```diff
@@ -571,9 +571,14 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           resultExpressions,
           planLater(child)))
 
-      case PhysicalAggregation(_, _, _, _) =>
+      case PhysicalAggregation(_, aggExpressions, _, _) =>
+        var groupAggPandasUDFNames = Set[String]()
+        aggExpressions.foreach(expr => if (expr.isInstanceOf[PythonUDF]) {
+          groupAggPandasUDFNames += expr.asInstanceOf[PythonUDF].name
+        })
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:578]: can we simplify this logic by, for example, aggExpressions.filter().map()?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 3]
```diff
@@ -571,9 +571,14 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
         // If cannot match the two cases above, then it's an error
-        throw QueryCompilationErrors.cannotUseMixtureOfAggFunctionAndGroupAggPandasUDFError()
+        throw QueryCompilationErrors.cannotUseMixtureOfAggFunctionAndGroupAggPandasUDFError(
+          "<" + groupAggPandasUDFNames.mkString(",") + ">")
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:581]: Can we use string interpolation here? BTW, is `<...>` format is consistent in the error message?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:581]: I think we should add a space in `mkString` like `mkString(", ")`

**Comment by HyukjinKwon** [comment]
There's another test failure:

```
[info] SparkThrowableSuite:
[info] - No duplicate error classes (5 milliseconds)
[info] - Error classes are correctly formatted *** FAILED *** (11 milliseconds)
[info]   "... %s.\n%s" ]
[info]     },
[info]     "[CAST_CAUSES_OVERFLOW" : {
[info]       "message" : [ "Casting %s to %s causes overflow. To return NULL instead, use 'try_cast'. If necessary set %s to false to bypass this error." ],
[info]       "sqlState" : "22005"
[info]     },
[info]     "CONCURRENT_QUERY" : {
[info]       "message" : [ "Another instance of this query was just started by a concurrent session." ]
[info]     },
[info]     "DATETIME_OVERFLOW" : {
[info]       "message" : [ "Datetime operation overflow: %s." ],
[info]       "sqlState" : "22008"
```

Mind fixing this togheter?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 3 above)
Comment (itholic) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:581]: Seems like we usually use single quotes for emphasize the name. Let me change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 4]
```diff
@@ -1986,7 +1987,7 @@ object QueryCompilationErrors {
     new AnalysisException(
       s"Failed to execute SHOW CREATE TABLE against table/view ${table.identifier}, " +
         "which is created by Hive and uses the following unsupported feature(s)\n" +
-        features.map(" - " + _).mkString("\n"))
+          features.map(" - " + _).mkString("\n"))
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1990]: Can we revert this change back?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 5]
```diff
@@ -571,9 +571,13 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           resultExpressions,
           planLater(child)))
 
-      case PhysicalAggregation(_, _, _, _) =>
+      case PhysicalAggregation(_, aggExpressions, _, _) =>
+        var groupAggPandasUDFNames = Set[String]()
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:575]: Can we directly assign this without using `var`?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:575]: Also, I think we should add a default string in case `groupAggPandasUDFNames` is empty to make the code future-proof.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 6]
```diff
@@ -1337,10 +1337,11 @@ object QueryCompilationErrors {
       "Stream-stream join without equality predicate is not supported", plan = Some(plan))
   }
 
-  def cannotUseMixtureOfAggFunctionAndGroupAggPandasUDFError(): Throwable = {
+  def cannotUseMixtureOfAggFunctionAndGroupAggPandasUDFError(
+      groupAggPandasUDFNames: Set[String]): Throwable = {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1341]: Can we pass `Seq[String]` instead?

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1341]: You can make it deduplicated by calling `distinct` method without converting it to `Set`.

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala [hunk 7]
```diff
@@ -111,14 +111,17 @@ class QueryCompilationErrorsSuite extends QueryTest with SharedSparkSession {
...
-        .agg(pandasTestUDF(df("Quantity")), sum(df("Quantity"))).collect()
+        .agg(pandasTestUDF1(df("Quantity")), pandasTestUDF2(df("Quantity")), sum(df("Quantity")))
+        .collect()
     }
 
-    assert(e.errorClass === Some("CANNOT_USE_MIXTURE"))
+    assert(e.errorClass === Some("INVALID_PANDAS_UDF_PLACEMENT"))
     assert(e.message ===
-      "Cannot use a mixture of aggregate function and group aggregate pandas UDF")
+      "The group aggregate pandas UDF 'pandas_udf_1, pandas_udf_2' cannot be invoked " +
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala:123]: If we're going to quote the function name, I think the output should look like `'pandas_udf_1', 'pandas_udf_2'`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 8]
```diff
@@ -1337,10 +1337,15 @@ object QueryCompilationErrors {
...
+      groupAggPandasUDFNames: Seq[String]): Throwable = {
     new AnalysisException(
-      errorClass = "CANNOT_USE_MIXTURE",
-      messageParameters = Array.empty)
+      errorClass = "INVALID_PANDAS_UDF_PLACEMENT",
+      messageParameters = if (groupAggPandasUDFNames.isEmpty) {
+        Array.empty
+      } else {
+        Array(groupAggPandasUDFNames.map(name => s"'$name'").mkString(", "))
+      })
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1348]:
```suggestion
      messageParameters = Array(groupAggPandasUDFNames.map(name => s"'$name'").mkString(", "))
```

**Review by HyukjinKwon — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 9]
```diff
@@ -1337,10 +1337,11 @@ object QueryCompilationErrors {
...
   }
 
-  def cannotUseMixtureOfAggFunctionAndGroupAggPandasUDFError(): Throwable = {
+  def cannotUseMixtureOfAggFunctionAndGroupAggPandasUDFError(
+      groupAggPandasUDFNames: Seq[String]): Throwable = {
     new AnalysisException(
-      errorClass = "CANNOT_USE_MIXTURE",
-      messageParameters = Array.empty)
+      errorClass = "INVALID_PANDAS_UDF_PLACEMENT",
+      messageParameters = Array(groupAggPandasUDFNames.map(name => s"'$name'").mkString(", "))
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1344]:
```suggestion
      messageParameters = Array(groupAggPandasUDFNames.map(name => s"'$name'").mkString(", ")))
```

**Comment by HyukjinKwon** [comment]
@itholic mind rebasing this please?

**Comment by itholic** [comment]
Just rebased!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 10]
```diff
@@ -1338,10 +1338,11 @@ object QueryCompilationErrors {
       "Stream-stream join without equality predicate is not supported", plan = Some(plan))
   }
 
-  def cannotUseMixtureOfAggFunctionAndGroupAggPandasUDFError(): Throwable = {
+  def cannotUseMixtureOfAggFunctionAndGroupAggPandasUDFError(
```
[the remark this answers is not in the document]
Comment (itholic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1341]: Sounds good. Let me address

**Comment by HyukjinKwon** [comment]
@itholic seems like pandas UDF tests fail. Can you take a look please?

```
======================================================================
ERROR [0.471s]: test_invalid_args (pyspark.sql.tests.test_pandas_udf_grouped_agg.GroupedAggPandasUDFTests)
----------------------------------------------------------------------
pyspark.sql.utils.AnalysisException: The group aggregate pandas UDF 'avg' cannot be invoked together with as other, non-pandas aggregate functions.

JVM stacktrace:
org.apache.spark.sql.AnalysisException: The group aggregate pandas UDF 'avg' cannot be invoked together with as other, non-pandas aggregate functions.
	at org.apache.spark.sql.errors.QueryCompilationErrors$.invalidPandasUDFPlacementError(QueryCompilationErrors.scala:[1345](https://github.com/itholic/spark/runs/6000818149?check_suite_focus=true#step:10:1345))
	at org.apache.spark.sql.execution.SparkStrategies$Aggregation$.apply(SparkStrategies.scala:579)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.$anonfun$plan$1(QueryPlanner.scala:63)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:486)
...

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/__w/spark/spark/python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py", line 491, in test_invalid_args
    df.groupby(df.id).agg(mean_udf(df.v), mean(df.v)).collect()
AssertionError: "mixture.*aggregate function.*group aggregate pandas UDF" does not match "The group aggregate pandas UDF 'avg' cannot be invoked together with as other, non-pandas aggregate functions.
```

**Comment by HyukjinKwon** [comment]
@itholic mind fixing the linter failure?

```
Run PYTHON_EXECUTABLE=python3.9 ./dev/lint-python
starting python compilation test...
python compilation succeeded.

starting black test...
black checks failed:
would reformat python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py
Oh no! 💥 💔 💥
1 file would be reformatted, 347 files would be left unchanged.
Please run 'dev/reformat-python' script.
1
```

File: python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py [hunk 11]
```diff
@@ -486,7 +486,9 @@ def test_invalid_args(self):
 
         with QuietTest(self.sc):
             with self.assertRaisesRegex(
-                AnalysisException, "mixture.*aggregate function.*group aggregate pandas UDF"
+                AnalysisException,
+                "The group aggregate pandas UDF 'avg' cannot be invoked together with as other, " +
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf_grouped_agg.py:490]: `+` at the end wouldn't be necessary.

**Comment by itholic** [comment]
Seems like the test is passed in my forked repo: https://github.com/itholic/spark/actions/runs/2160323293

**Comment by HyukjinKwon** [comment]
Merged to master.
